### PR TITLE
Set nsPagedSizeLimit=20000 on creation of pkidbuser

### PR DESCRIPTION
**Pull request by ckelleyRH**
Currently every entry is pulled by default which has been causing users problems. There is an RFE to change this attribute dynamically based on the size of the backend DB but for now set a reasonable default.

Resolves: #1959057

### CODE REVIEW

**Comment by ckelleyRH** [comment]
I've not verified that `pkidbuser` has the additional attribute with an LDAP search or anything yet but figured I would put the patch for review anyway so you guys can see if I'm on the right path.

File: base/server/python/pki/server/deployment/__init__.py
Comment (edewata) [inline base/server/python/pki/server/deployment/__init__.py:2582]: The `add_see_also` and `del_see_also` params were provided to add/remove a `seeAlso` attribute (which is multi-valued) in LDAP:
https://github.com/dogtagpki/pki/blob/master/base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java#L790-L793
https://github.com/dogtagpki/pki/blob/master/base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java#L832-L835

To set an `nsPagedSizeLimit` attribute (which is single-valued) I'd suggest modifying `UGSubsystem` to do something like this:
```
LDAPModificationSet mods = new LDAPModificationSet();

if (seeAlsoToAdd != null) {
    LDAPAttribute attr = new LDAPAttribute("seeAlso");
    attr.addValue(seeAlsoToAdd);
    mods.add(LDAPModification.ADD, attr);
}

if (seeAlsoToRemove != null) {
    LDAPAttribute attr = new LDAPAttribute("seeAlso");
    attr.addValue(seeAlsoToRemove);
    mods.add(LDAPModification.DELETE, attr);
}

if (pagedSizeLimit != null) {
    LDAPAttribute attr = new LDAPAttribute("nsPagedSizeLimit");
    attr.addValue(pagedSizeLimit);
    mods.add(LDAPModification.REPLACE, attr);
}

ldapconn.modify(dn, mods);
```
So I think it will require some non-trivial changes, but it shouldn't be too complicated.

Comment (edewata) [reply base/server/python/pki/server/deployment/__init__.py:2582]: Also, this code is currently not tested. You might want to modify an existing test to verify that the `nsPagedSizeLimit` attribute was indeed set in LDAP.

Comment (edewata) [reply base/server/python/pki/server/deployment/__init__.py:2582]: Another thing is, this change will affect new installation only. The existing `pkidbuser` will not be updated and we currently don't have a mechanism to upgrade the database.
https://github.com/dogtagpki/pki/issues/1279
https://github.com/dogtagpki/pki/issues/1476

Also, since this issue is mainly affecting IPA, this upgrade probably can be done by IPA since IIUC they have a mechanism for that.

Comment (fmarco76) [reply base/server/python/pki/server/deployment/__init__.py:2582]: > I'd suggest modifying `UGSubsystem` to do something like this

If the user is created in the UGSubSystem then the new size limit can be set during the creation of the user taking the value from the ldap configuration. The dynamic change based on the real size of the entries need some additional thought, maybe can be added to the pruning job which could have an estimation of the entries size.

Comment (edewata) [reply base/server/python/pki/server/deployment/__init__.py:2582]: The current pruning job removes expired entries at a fixed rate regardless of database size. I'm not sure if DS provides a way to get the number of entries in the database. Maybe it's possible with VLV, but in the future we might drop VLV, so we shouldn't rely on that.

Comment (fmarco76) [reply base/server/python/pki/server/deployment/__init__.py:2582]: If it is not possible to get the number of remaining entry a different approach is needed. That was just on option but we can create a separate process/thread/cron.

Comment (ckelleyRH) [reply base/server/python/pki/server/deployment/__init__.py:2582]: Thanks for the feedback! I will take a look at modifying `UGSubsystem` and adding a test. I don't think the upgrade is necessary as the problem is easy to spot and the workaround is simple - just an LDAP modify command (which has been tested in the field by support).

Comment (ckelleyRH) [reply base/server/python/pki/server/deployment/__init__.py:2582]: As this is going to simply add a static value, would it be reasonable to modify `addUser` to put `nsPagedSizeLimit` into the existing `LDAPAttributeSet`?

Comment (edewata) [reply base/server/python/pki/server/deployment/__init__.py:2582]: That should be fine, but be sure to confirm with IPA/support team that this solution will be sufficient for them, and there will be no automatic upgrade for existing instances.

File: .github/workflows/ca-existing-ds-test.yml
Comment (fmarco76) [inline .github/workflows/ca-existing-ds-test.yml:257]: Maybe, we could add test with the expected value.

**Review by fmarco76 — APPROVED** [review]
LGTM, just a comment for the test but feel free to merge.

File: .github/workflows/ca-existing-ds-test.yml
Comment (ckelleyRH) [reply .github/workflows/ca-existing-ds-test.yml:257]: I didn't want to make it too fragile, someone may want to add additional users (or remove one) and that would break this. Actually, the `wc -l` part was a leftover from something I was trying but it is not necessary. `grep` returns 1 if there is no match and that is sufficient for us, I think. I will drop `wc -l` before merging as that is returning 0 and probably hiding the error code from `grep`.

Comment (ckelleyRH) [reply .github/workflows/ca-existing-ds-test.yml:257]: https://github.com/dogtagpki/pki/actions/runs/4064058832/jobs/6997310640
Turns out `wc -l` was hiding `grep`s return code of 1! Not quite sure why, perhaps I needed to make the `-b` argument the same as the other tests?

Comment (edewata) [reply .github/workflows/ca-existing-ds-test.yml:257]: How about something like this?
```
echo "dn: uid=pkidbuser,ou=People,dc=ca,dc=pki,dc=example,dc=com" > expected
echo "nsPagedSizeLimit: 20000" >> expected

docker exec ds ldapsearch \
    -H ldap://ds.example.com:3389 \
    -D "cn=Directory Manager" \
    -w Secret.123 \
    -b "uid=pkidbuser,ou=People,dc=ca,dc=pki,dc=example,dc=com" \
    nsPagedSizeLimit \
    | tee actual

diff expected actual
```
This way we test `pkidbuser` specifically, the expected value is clearly defined, and if there's a problem we can see the actual value in the log.

Comment (ckelleyRH) [reply .github/workflows/ca-existing-ds-test.yml:262]: OK, this might be what @fmarco76 was alluding to before, sorry if I misunderstood you!

Out of interest, why is the port 3389 here rather than 389? Is this a container related thing?

File: base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java
Comment (edewata) [inline base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java:704]: Actually the `nsPagedSizeLimit` attribute only needs to be added to an LDAP user that will be used to create LDAP connections. In IPA case that would be `pkidbuser`. However, for general PKI users (e.g. admins, agents) this attribute is not needed, so it should be optional. Also, in case we need to change the value later it will be challenging to update all users in LDAP, so I think we should limit that to `pkidbuser` only.

I'd suggest we change [this code](https://github.com/dogtagpki/pki/blob/master/base/server/python/pki/server/deployment/__init__.py#L2644-L2648) to specifically add the attribute to `pkidbuser` like this:
```
subsystem.add_user(
    'pkidbuser',
    full_name='pkidbuser',
    user_type='agentType',
    state='1',
    nsPagedSizeLimit=20000)
```
or this if you want to make it more reusable:
```
subsystem.add_user(
    'pkidbuser',
    full_name='pkidbuser',
    user_type='agentType',
    state='1',
    attributes={
        nsPagedSizeLimit: 20000
    })
```

Comment (ckelleyRH) [reply base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java:704]: There's no code for `attributes` right, so I would need to add a new argument for `add_user` and have code in the Java side to handle that new field?

Comment (edewata) [reply base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java:704]: That's correct.

File: base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java
Comment (ckelleyRH) [inline base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java:80]: I will remove this when I fix the broken test, but just pushing solution for now to see if the CI likes it (and the reviewers, of course :+1: )

**Comment by ckelleyRH** [comment]
The new method accepts multiple attributes but the current test doesn't verify this (I have verified it on my machine though). Do you think it is worth writing a new test for this @edewata @fmarco76?

File: base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java
Comment (edewata) [inline base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java:708]: The `attributes.size() > 0` check might be redundant since the `for` loop will not do anything if `attributes` is empty.

File: base/server/python/pki/server/deployment/__init__.py
Comment (edewata) [inline base/server/python/pki/server/deployment/__init__.py:2654]: I think Python has a preference for single-quotes, although double-quotes are acceptable too.

File: base/server/python/pki/server/subsystem.py
Comment (edewata) [inline base/server/python/pki/server/subsystem.py:1645]: Same thing about single vs. double-quotes.

File: base/server/src/main/java/com/netscape/cmscore/usrgrp/User.java
Comment (edewata) [inline base/server/src/main/java/com/netscape/cmscore/usrgrp/User.java:108]: The `m` prefix for field names is a legacy convention. For new code I don't think we should be using that anymore. Just use a normal Java naming convention.

Also, the `null` initialization is redundant since non-primitive fields are initialized to `null` by default.

File: docs/changes/v11.4.0/Server-Changes.adoc
Comment (edewata) [inline docs/changes/v11.4.0/Server-Changes.adoc:7]: I'd suggest showing the entire `pki-server <subsystem>-add-user` command instead of just the value.

**Review by edewata — APPROVED** [review]
Thanks for the update! I have some minor comments below, but I think the code in general looks good. The current test is probabl sufficient since it will exercise the new `--attributes` param internally. Feel free to update/merge.

File: .github/workflows/ca-existing-ds-test.yml
Comment (edewata) [reply .github/workflows/ca-existing-ds-test.yml:262]: I think you can add `-o ldif_wrap=no` and `-LLL` to the `ldapsearch` command to normalize and clean up the output for verification.

File: base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java
Comment (ckelleyRH) [reply base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java:708]: Ack, I cargo-culted the bit for the TPS profiles. I have changed this anyway in next revision to use `isEmpty()` so I might as well just drop it.

File: base/server/python/pki/server/subsystem.py
Comment (ckelleyRH) [reply base/server/python/pki/server/subsystem.py:1645]: Ack.

File: base/server/src/main/java/com/netscape/cmscore/usrgrp/User.java
Comment (ckelleyRH) [reply base/server/src/main/java/com/netscape/cmscore/usrgrp/User.java:108]: Yeah there was a lot here I didn't particularly like, I originally used a more modern convention but it looked ugly with two different styles so I kept it like the original code. I will leave this like it is so the diff for this change is minimised, then push a PR to clean this class up.

File: docs/changes/v11.4.0/Server-Changes.adoc
Comment (ckelleyRH) [reply docs/changes/v11.4.0/Server-Changes.adoc:7]: Ack.

File: .github/workflows/ca-existing-ds-test.yml
Comment (ckelleyRH) [reply .github/workflows/ca-existing-ds-test.yml:262]: Thanks, will be useful. I'll merge once I finally make this test pass :laughing:

File: base/server/python/pki/server/deployment/__init__.py
Comment (edewata) [reply base/server/python/pki/server/deployment/__init__.py]: This is what I saw earlier, but I'm not sure if this is standard or widely accepted:
https://docs.ckan.org/en/ckan-2.1.5/python-coding-standards.html

**Comment by ckelleyRH** [comment]
Thanks for all the input @fmarco76 @edewata !

**Pull request merged**
